torch_geometric/explain/metrics/faithfulness.py

```diff
@@ -0,0 +1,169 @@
+  code = {https://github.com/mims-harvard/GraphXAI},
+  publisher = {arXiv},
+  year = {2022},
+  copyright = {Creative Commons Attribution 4.0 International}
+
+}
+
+
+we obtain the prediction probability vector ŷ_u using the GNN, i.e. ŷ_u = f (S_u),
+and using the explanation, i.e. ŷ_u′ = f (S_u′), where we generate a masked subgraph S_u′ by only keeping the original
+values of the top-k features identified by an explanation, and get their respective predictions ŷ_u′
+
+Un-Faithfulness of a GNN predictor is given as
+GEF(f,S_u,S_u') = 1 - exp^-KL(ŷ_u || ŷ_u')
+
+Note that higher values indicate a higher degree of unfaithfulness
+"""
+
+import torch
+import torch.nn.functional as F
+from utils import perturb_node_features
+
+from torch_geometric.data import Data
+from torch_geometric.explain import Explanation
+
+
+def get_node_faithfulness(node_idx: int, data: Data, explanation: Explanation,
+                          model: torch.nn.Module, topk: int, device: str,
+                          **kwargs):
+    '''
+    This function should be called for node level tasks to calculate the node's faitfulness score
+
+    Args:
+        node_idx (int): node_idx for node to explain, should be the same used for explainer (it can be get from explainer)
+        data (Data) : dataset
+        explanation (Explanation): Explanation output generated by an explainer
+        model (torch.nn.Module): model used in explanation (ait can be get from explanation)
+        topk (int): k most important feature/nodes, it does not calculate percentage and
+                    the user should take into account the threshold config used in explanation
+        device (str): cpu / cuda
+        kwargs (dict, optional): Any additional arguments to the forward method
+            of model, other than x and edge_index.
+    '''
+    node_faithfulness = node_feat_faithfulness = None
+
+    try:
+        with torch.no_grad():
+            org_vec = model(data.x, data.edge_index, **kwargs)[node_idx]
+            org_softmax = F.softmax(org_vec, dim=-1)
+    except:
+        raise Exception(
+            "Not able to get get original predictions from model and data")
+
+    if getattr(explanation, 'node_mask', None) is not None:
+        if topk > explanation.node_mask.shape[0]:
+            raise ValueError(
+                "Topk cannot be greater than nodes size, take also in consideration threshold config in explainer"
+            )
+
+        # getting top k important nodes according to explanation
+        top_k_nodes = explanation.node_mask.topk(topk)[1]
+
+        pert_x = data.x.clone()
+        # removing not top-k
+        rem_nodes = [
+            node for node in range(data.x.shape[0]) if node not in top_k_nodes
+        ]
+        pert_x[rem_nodes] = torch.zeros_like(pert_x[rem_nodes]).to(device)
+        # getting predictions of explainer
+        pert_vec = model(pert_x, data.edge_index, **kwargs)[node_idx]
```

Comment:
Can we reuse `Expalnation.get_explanation_subgraph`. Also here you only seem to be using `node_mask` but we should be using the edge_mask too right?

Reply:
I tired to use it but I did not work I do not remember why :) I will check it again tell the reason. According to definition of faithfulness, it is calculated by masking nodes or by perturbation of node features. I am not an expert on explanaibility so maybe we should discuss this in detail with the issue owners.

Reply:
Issue refers to [Evaluating Explainability for Graph Neural Networks](https://arxiv.org/abs/2208.09339) but it is based on the paper you linked.  I have also read this paper. They mask edges or rewire edges with some probability pr. I will try to add this.

Reply:
For node level tasks if I use get_explanation_subgraph, I loose node_idx as there is no mapping. Maybe I am wrong!

Reply:
I added a function with subgraphs but not really sure. It was just to discuss. It corresponds well to explanation on the paper : 

1. get model’s prediction for node u ŷ_u = f(S_u)
2. generate a masked subgraph S_u′ by only keeping the original values of the top-k features identified by an explanation
3. get model's prediction for S_u' 
4. calculate metric 

Explanation gets subgraph with different masks for me but I think I need to use node_idx to get the prediction for node u and there is no mapping returned. I am a kind of lost here :)

Reply:
Oh i see the problem you are facing `node_idx` gets renumbered when you call `Explanation.get_explanation_subgraph`. In that case i think its best not to use `Explanation.get_explanation_subgraph`. 
Another thing you could try is to get the new node id
```
# The node being explained should not be masked out
assert explanation.node_maks[node_idx]>0 
node_idx = (explanation.node_mask>0).cumsum(dim=0)[node_idx]
```

Reply:
Yepp that was my problem since the beginning. I did not know we could get it this way, this is the expert touch :) Thanks

`sub_node_idx = (explanation.node_mask>0).cumsum(dim=0)[node_idx]`
In that case I believe , `explanation.x[node_idx]` should be equal to `explanation.get_explanation_subgraph().x[sub_node_idx]`
but it is not. When I check all the nodes in subgraph to find corresponding `explanation.x[node_idx]`, i see that corresponding sub_node_idx is always 0.

Reply:
`sub_node_idx = (explanation.node_mask>0).cumsum(dim=0)[node_idx]`
In that case I believe , `explanation.x[node_idx]` should be equal to `explanation.get_explanation_subgraph().x[sub_node_idx]`
but it is not. When I check all the nodes in subgraph to find corresponding `explanation.x[node_idx]`, i see that corresponding sub graph node_idx is always sub_node_idx-1. What am I missing?

Reply:
Yeah that makes sense it should be `sub_node_idx-1`, since the tensors are 0 indexed.

Reply:
@ZeynepP This will also be included in the revamp of the `Explanation` class, so don't worry about this for now 😃